planning/motion_velocity_planner/autoware_motion_velocity_obstacle_stop_module/src/obstacle_stop_module.cpp

```diff
@@ -897,10 +957,44 @@ double ObstacleStopModule::calc_desired_stop_margin(
 {
   // calculate default stop margin
   const double default_stop_margin = [&]() {
+    const double v_ego = planner_data->current_odometry.twist.twist.linear.x;
+    const double v_obs = stop_obstacle.velocity;
+
     const auto ref_traj_length =
       autoware::motion_utils::calcSignedArcLength(traj_points, 0, traj_points.size() - 1);
-    if (dist_to_collide_on_ref_traj > ref_traj_length) {
-      // Use terminal margin (terminal_stop_margin) for obstacle stop
+    if (v_obs < stop_planning_param_.max_negative_velocity) {
+      const double a_ego = stop_planning_param_.effective_deceleration_opposing_traffic;
+      const double & bumper_to_bumper_distance = stop_obstacle.dist_to_collide_on_decimated_traj;
+
+      const double braking_distance = v_ego * v_ego / (2 * a_ego);
+      const double stopping_time = v_ego / a_ego;
+      const double distance_obs_ego_braking = std::abs(v_obs * stopping_time);
+
+      const double ego_stop_margin = stop_planning_param_.stop_margin_opposing_traffic;
+
+      const double rel_vel = v_ego - v_obs;
+      const double epsilon = 1e-6;  // Small threshold for numerical stability
+      if (std::abs(rel_vel) <= epsilon) {
+        RCLCPP_WARN(
+          logger_,
+          "Relative velocity (%.3f) is too close to zero. Using minimum safe value for "
+          "calculation.",
+          rel_vel);
+        return stop_planning_param_.stop_margin;  // Return default stop margin as fallback
+      }
+
+      const double T_coast = std::max(
+        (bumper_to_bumper_distance - ego_stop_margin - braking_distance +
+         distance_obs_ego_braking) /
+          rel_vel,
+        0.0);
+
+      const double stopping_distance = v_ego * T_coast + braking_distance;
+
+      const double stop_margin = bumper_to_bumper_distance - stopping_distance;
+
+      return stop_margin;
+    } else if (dist_to_collide_on_ref_traj > ref_traj_length) {
```

Comment:
```suggestion
    }
    if (dist_to_collide_on_ref_traj > ref_traj_length) {
```
Please use early return.

Reply:
Done.